spark/spark-scala-parent/src/main/scala/org/apache/zeppelin/spark/BaseSparkScalaInterpreter.scala

```diff
@@ -79,7 +79,35 @@ abstract class BaseSparkScalaInterpreter(val conf: SparkConf,
     System.setProperty("scala.repl.name.line", ("$line" + this.hashCode).replace('-', '0'))
   }
 
-  protected def interpret(code: String, context: InterpreterContext): InterpreterResult
+  def interpret(code: String, context: InterpreterContext): InterpreterResult = {
+    if (context != null) {
+      interpreterOutput.setInterpreterOutput(context.out)
+      context.out.clear()
+    }
+
+    Console.withOut(if (context != null) context.out else Console.out) {
+      interpreterOutput.ignoreLeadingNewLinesFromScalaReporter()
+      val lastStatus = scalaInterpret(code) match {
+        case scala.tools.nsc.interpreter.IR.Success =>
+          InterpreterResult.Code.SUCCESS
+        case error@scala.tools.nsc.interpreter.IR.Error =>
+          InterpreterResult.Code.ERROR
+        case scala.tools.nsc.interpreter.IR.Incomplete =>
+          // add print("") at the end in case the last line is comment which lead to INCOMPLETE
+          scalaInterpret(code + "\nprint(\"\")") match {
```

Comment:
looks like we are re-running (scalaInterpret) the code in this case? I'm not sure about this - since the code might not be idempotent?

Reply:
It won't rerun the code,  see the test here. https://github.com/apache/zeppelin/pull/3055/files#diff-b3caeb6afe474aec6d5352ac3ee3a02aR114